### ISSUE TITLE
Develop token based API authentication

### ISSUE THREAD

**Issue by Connoropolous** [Hylozoic/hylo-node#98]
I have some ideas about how to move it along, based on recent work with Loomio and Metamaps, but just curious to hear any background first :)

**Comment by levity** [comment]
We haven't done anything beyond the KISS token. I imagine that we could just generate API tokens for users that would basically behave as second passwords, like many other services do. Depending on user needs, we could specify for each token whether it is read-only or read-write.

**Comment by Connoropolous** [comment]
Cool. I have just completed a very short sprint of work on this with Metamaps, so I have some things fresh on my mind. 

First of all, I imagine doing first iteration of it with no UI. Or the UI/UX could be developed in parallel, but I doubt @themadray has bandwidth for it right now cause of the main work he's doing. 
I think enabling read-only would also be a second phase of work, but the first phase should definitely account for that happening in the not too distant future. I am just thinking that we prbly don't have a use case for it right away. Enabling token auth to behave exactly like a second password, as you mention, is where my thinking is at, at the moment. 

In metamaps, I created three endpoints, which should be sufficient for the UI when it gets built as well.
CREATE, MY_TOKENS, DESTROY

Create accepts a `description` param only, and generates the token on the server and sends it back. 
My_tokens, if you are authed, returns an array of JSON objects of your tokens
Destroy only requires the id of the token you want deleted to be specified, and verifies that you are its owner. 

So in the implementation I just did, we have the Token model looking like this: 
https://github.com/metamaps/metamaps_gen002/blob/develop/app/models/token.rb
And the schema looking like this: 
https://github.com/metamaps/metamaps_gen002/blob/3aec00e07cfe97eb3616d157148b422195671de4/db/migrate/20160310200131_create_tokens.rb

Are there concerns you would have about a similar implementation @levity ? Further refinements for the Hylo use case? I am looking to actually go forward on implementation of this, so I am trying to get as good an idea ahead of time about design for it, so skip a few rounds of feedback after the fact.
